vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/request_for.h

```diff
@@ -0,0 +1,56 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_ENDPOINTS_REQUEST_FOR_H_
+#define BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_ENDPOINTS_REQUEST_FOR_H_
+
+#include <type_traits>
+#include <utility>
+
+#include "base/callback.h"
+#include "bat/ledger/internal/endpoints/request_builder.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/logging/logging.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace ledger::endpoints {
+
+template <typename Endpoint>
+class RequestFor {
+ public:
+  template <typename... Ts>
+  RequestFor(LedgerImpl* ledger, Ts&&... ts)
+      : ledger_(ledger),
+        request_(Endpoint{ledger, std::forward<Ts>(ts)...}.Request()) {
+    static_assert(std::is_base_of_v<RequestBuilder, Endpoint>,
+                  "Endpoint should be derived from RequestBuilder!");
+    DCHECK(ledger_);
+  }
+
+  RequestFor(const RequestFor&) = delete;
+  RequestFor& operator=(const RequestFor&) = delete;
+
+  RequestFor(RequestFor&&) = delete;
+  RequestFor& operator=(RequestFor&&) = delete;
+
+  void Send(base::OnceCallback<void(typename Endpoint::Result&&)> callback) && {
+    if (!request_ || !*request_) {
+      BLOG(0, "Failed to create request!");
+      return std::move(callback).Run(
+          base::unexpected(Endpoint::Error::FAILED_TO_CREATE_REQUEST));
```

Comment:
I'm wondering if there's another option besides a generic "failed to create request" error that's duplicated across all result enums.

The other option that I can think of is to have the RequestBuilder methods go back to returning `base::expected`, but with an `Error` as the error type.

```cpp
base::expected<std::string, MyEndpointError> Content() override {
  return base::unexpected(MyEndpointError::kInvalidRewardsWallet);
}
```

But then `RequestBuilder` would also need to be CRTP.

It seems like it's fine as-is for now though.

Reply:
We'd end up having several interface classes (`RequestBuilder<>`, one per endpoint) instead of just one (`RequestBuilder`), since e.g. `base::expected<std::string, MyEndpointError1>` and `base::expected<std::string, MyEndpointError2>` are neither the same type, nor covariant types. That's just a side effect, but we'd also run into a "chicken or the egg" problem:
```
template <typename Endpoint>
class RequestBuilder {
 public:
// ...
  virtual base::expected<std::string, /* >>> */ typename Endpoint::Error /* <<< */> Content();
// ...
};
```
At that point `Endpoint` is an incomplete type, so you can't use its nested typedef.
  There would be workarounds (e.g. passing `Error` as `RequestBuilder<>`'s template parameter, using a trait class, building a comple-time map and do a lookup while `RequestBuilder<>` is instantiated, etc.), but neither is optimal, unfortunately.